### PR TITLE
Fix REPO-2903 Service Pack: MNT-18285: Brute force account attacks: Accounts not reenabled when using authentication chaining

### PR DESCRIPTION
**Pull request by cristianturlica**
- Logins are now protected based on a combined key from authentication service system id and user name, this allows us to fix the case when a valid login was denied for subsequent authentication service if the prior authentication services in the chain failed.
- Cache is now set to be local, as the new implementation doesn't require it to be fully distributed

### CODE REVIEW

File: src/main/java/org/alfresco/repo/security/authentication/AuthenticationServiceImpl.java [hunk 1]
```diff
@@ -42,7 +43,10 @@
     private Log logger = LogFactory.getLog(AuthenticationServiceImpl.class);
     AuthenticationComponent authenticationComponent;
     TicketComponent ticketComponent;
-    
+
+    /** a uuid identifying this unique instance */
+    private String uuid;
```
Comment (derekhulley) [inline src/main/java/org/alfresco/repo/security/authentication/AuthenticationServiceImpl.java]: The variable name does not really describe what it does.
Please use something like: `serviceInstanceId`

Comment (cristianturlica) [reply src/main/java/org/alfresco/repo/security/authentication/AuthenticationServiceImpl.java]: I will update it, thanks.

File: src/main/java/org/alfresco/repo/security/authentication/AuthenticationServiceImpl.java [hunk 2]
```diff
@@ -186,10 +195,15 @@ public void recordFailedAuthentication(String userName)
                     }
                 }
             }
-            protectedUsersCache.put(userName, protectedUser);
+            protectedUsersCache.put(protectedUserKey, protectedUser);
         }
     }
 
+    private String getProtectedUserKey(String userName)
```
Comment (derekhulley) [inline src/main/java/org/alfresco/repo/security/authentication/AuthenticationServiceImpl.java]: Javadoc to say what the method does.  It's private so there is no way to guess that it combines the service instance ID with the username.

Comment (cristianturlica) [reply src/main/java/org/alfresco/repo/security/authentication/AuthenticationServiceImpl.java]: I will make it public and add the java doc.

File: src/main/java/org/alfresco/repo/security/authentication/AuthenticationServiceImpl.java [hunk 3]
```diff
@@ -186,10 +195,15 @@ public void recordFailedAuthentication(String userName)
...
+    {
+        return uuid + userName;
```
Comment (derekhulley) [inline src/main/java/org/alfresco/repo/security/authentication/AuthenticationServiceImpl.java]: `return uuid + userName;`
`return uuid + "@@" + userName;`

**Comment by derekhulley** [comment]
None of the commits actually say how the fix was achieved.
Please add it to the merge commit so that we can know what was done to fix the issue without looking at the code.

File: src/main/java/org/alfresco/repo/security/authentication/AuthenticationServiceImpl.java (hunk 2 above)
Comment (derekhulley) [reply src/main/java/org/alfresco/repo/security/authentication/AuthenticationServiceImpl.java]: Private is fine.  But it needs to be commented.

File: src/test/java/org/alfresco/repo/security/authentication/AuthenticationServiceImplTest.java [hunk 4]
```diff
@@ -57,6 +58,7 @@
     private SimpleCache<String, ProtectedUser> cache;
     private TicketComponent ticketComponent = mock(TicketComponent.class);
     private AuthenticationServiceImpl authService;
+    private AuthenticationServiceImpl authService2;
```
Comment (mattward) [inline src/test/java/org/alfresco/repo/security/authentication/AuthenticationServiceImplTest.java]: Would be nice to have a comment why there is a second instance of the ASI class

File: src/main/java/org/alfresco/repo/security/authentication/AuthenticationServiceImpl.java (hunk 3 above)
Comment (cristianturlica) [reply src/main/java/org/alfresco/repo/security/authentication/AuthenticationServiceImpl.java]: Done

**Pull request merged**